stdlib/asm/crypto/hashes/blake3.masm

```diff
@@ -1,579 +1,447 @@
-    drop
-    pushw.mem           # = c
-
-    pushw.local.0
-    drop
-    repeat.2
-        swap
-        drop
-    end
-    pushw.mem           # = b
-
-    pushw.local.0
-    repeat.3
-        swap
-        drop
-    end
-    pushw.mem           # = a
-end
-
-# initial hash state of blake3 when computing 2-to-1 hash i.e. two blake3 digests are being merged into single digest of 32 -bytes
-# see https://github.com/itzmeanjan/blake3/blob/f07d32ec10cbc8a10663b7e6539e0b1dab3e453b/include/blake3.hpp#L1709-L1713
-proc.initialize_hash_state.1
-    popw.local.0
-
-    # blake3 initial values
-    # see https://github.com/BLAKE3-team/BLAKE3/blob/da4c792d8094f35c05c41c9aeb5dfe4aa67ca1ac/reference_impl/reference_impl.rs#L36-L38
+# Initializes four memory addresses, provided for storing initial 4x4 blake3 
+# state matrix ( i.e. 16 elements each of 32 -bit ), for computing blake3 2-to-1 hash
+#
+# Expected stack state:
+#
+# [state_0_3_addr, state_4_7_addr, state_8_11_addr, state_12_15_addr]
+#
+# Note, state_`i`_`j`_addr -> absolute address of {state[i], state[i+1], state[i+2], state[i+3]} in memory | j = i+3
+#
+# Final stack state:
+#
+# [...]
+#
+# Initialized stack state is written back to provided memory addresses.
+#
+# Functionally this routine is equivalent to https://github.com/itzmeanjan/blake3/blob/f07d32e/include/blake3.hpp#L1709-L1713
+proc.initialize
     push.0xA54FF53A.0x3C6EF372.0xBB67AE85.0x6A09E667
-
-    pushw.local.0
-    repeat.3
-        swap
-        drop
-    end
-
+    movup.4
     popw.mem
 
     push.0x5BE0CD19.0x1F83D9AB.0x9B05688C.0x510E527F
-
-    pushw.local.0
-    drop
-    repeat.2
-        swap
-        drop
-    end
-
+    movup.4
     popw.mem
 
     push.0xA54FF53A.0x3C6EF372.0xBB67AE85.0x6A09E667
-
-    pushw.local.0
-    drop
-    drop
-    swap
-    drop
-
+    movup.4
     popw.mem
 
-    # blake3 hash constants https://github.com/itzmeanjan/blake3/blob/1c58f6a343baee52ba1fe7fc98bfb280b6d567da/include/blake3_consts.hpp#L16-L20
     push.11.64.0.0
-
-    pushw.local.0
-    drop
-    drop
-    drop
-
+    movup.4
     popw.mem
 end
 
-# permutes ordered message words, kept on stack top ( = sixteen 32 -bit BLAKE3 words )
-# such that next round of mixing can be applied
-# after completion of permutation, message words are transferred back to stack top, in ordered form
-# see https://github.com/itzmeanjan/blake3/blob/f07d32ec10cbc8a10663b7e6539e0b1dab3e453b/include/blake3.hpp#L1623-L1639
-proc.blake3_msg_words_permute.3
-    movup.2
-    movup.6
-    swap
-    movup.4
+# Permutes ordered message words, kept on stack top ( = sixteen 32 -bit BLAKE3 words )
+#
+# Expected stack top: 
+#
+# [s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10, s11, s12, s13, s14, s15]
+#
+# After permutation, stack top:
+#
+# [s2, s6, s3, s10, s7, s0, s4, s13, s1, s11, s12, s5, s9, s14, s15, s8]
+#
+# See https://github.com/itzmeanjan/blake3/blob/f07d32ec10cbc8a10663b7e6539e0b1dab3e453b/include/blake3.hpp#L1623-L1639
+proc.permute_msg_words
```

Comment:
Could you check if the following sequence works here? If so, it should be a bit more efficient:
```
movdn.7
movup.5
movdn.2
movup.4
movdn.7
swapw.3
swap
movdn.7
swapdw
movup.2
movdn.7
swapw
swapw.2
movup.3
movdn.6
movdn.5
movup.3
swapw
movup.3
swapdw
```

In general, this feels like a linear programming problem, and maybe at some point we can build a generic tool to find the optimal sequence of instructions which transforms state `a` into state `b`.

Reply:
This works as expected, thanks for the suggestion.